roachtest: cdc/crdb-chaos/rangefeed=true failed

SHA: https://github.com/cockroachdb/cockroach/commits/82026117d83262e87873aad52b8eca2dd0bea335

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1126329&tab=buildLog

```
The test failed on provisional_201902052032_v2.2.0-alpha.20190211:
	test.go:748,cdc.go:744,cdc.go:215,cdc.go:433: max latency was more than allowed: 21m15.099334977s vs 10m0s

```

SHA: https://github.com/cockroachdb/cockroach/commits/40e403544d60b1a44b8b1ed961a817c77d67aa31

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1131411&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:610,cdc.go:122,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1131411-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		   1.9      0.0      0.0      0.0      0.0 delivery
		  17m11s     3929            0.0           19.2      0.0      0.0      0.0      0.0 newOrder
		  17m11s     3929            0.0            1.9      0.0      0.0      0.0      0.0 orderStatus
		  17m11s     3929            0.0           18.9      0.0      0.0      0.0      0.0 payment
		  17m11s     3929            0.0            1.9      0.0      0.0      0.0      0.0 stockLevel
		E190209 07:30:05.445925 1 workload/cli/run.go:402  error in newOrder: dial tcp 10.142.0.109:26257: connect: connection refused
		  17m12s     3946            0.0            1.9      0.0      0.0      0.0      0.0 delivery
		  17m12s     3946            0.0           19.2      0.0      0.0      0.0      0.0 newOrder
		  17m12s     3946            0.0            1.9      0.0      0.0      0.0      0.0 orderStatus
		  17m12s     3946            0.0           18.9      0.0      0.0      0.0      0.0 payment
		  17m12s     3946            0.0            1.9      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1585,cdc.go:210,cdc.go:430,test.go:1206: unexpected node event: 2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/10f8010fa5778e740c057905e2d7664b5fd5d647

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1135549&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:612,cdc.go:124,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1135549-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		    1.8      0.0      0.0      0.0      0.0 delivery
		  22m35s     5044            0.0           18.6      0.0      0.0      0.0      0.0 newOrder
		  22m35s     5044            0.0            1.8      0.0      0.0      0.0      0.0 orderStatus
		  22m35s     5044            0.0           18.0      0.0      0.0      0.0      0.0 payment
		  22m35s     5044            0.0            1.8      0.0      0.0      0.0      0.0 stockLevel
		E190213 07:38:12.898689 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.187:26257: connect: connection refused
		  22m36s     5057            0.0            1.8      0.0      0.0      0.0      0.0 delivery
		  22m36s     5057            0.0           18.6      0.0      0.0      0.0      0.0 newOrder
		  22m36s     5057            0.0            1.8      0.0      0.0      0.0      0.0 orderStatus
		  22m36s     5057            0.0           18.0      0.0      0.0      0.0      0.0 payment
		  22m36s     5057            0.0            1.8      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1585,cdc.go:212,cdc.go:432,test.go:1212: unexpected node event: 2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/2e16b7357d5a15d87cd284d5f2c12e424ed29ca1

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1146277&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:612,cdc.go:124,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1146277-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		    1.7      0.0      0.0      0.0      0.0 delivery
		   6m23s     1074            0.0           18.0      0.0      0.0      0.0      0.0 newOrder
		   6m23s     1074            1.0            1.8      3.7      3.7      3.7      3.7 orderStatus
		   6m23s     1074            0.0           17.9      0.0      0.0      0.0      0.0 payment
		   6m23s     1074            0.0            1.8      0.0      0.0      0.0      0.0 stockLevel
		E190221 07:09:31.763372 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.134:26257: connect: connection refused
		   6m24s     1090            0.0            1.7      0.0      0.0      0.0      0.0 delivery
		   6m24s     1090            0.0           17.9      0.0      0.0      0.0      0.0 newOrder
		   6m24s     1090            0.0            1.8      0.0      0.0      0.0      0.0 orderStatus
		   6m24s     1090            0.0           17.9      0.0      0.0      0.0      0.0 payment
		   6m24s     1090            0.0            1.8      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1585,cdc.go:212,cdc.go:432,test.go:1212: unexpected node event: 2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/959dcf7de0f94cfcfa0062387b109adebd1f11da

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1163702&tab=buildLog

```
The test failed on master:
	cluster.go:1244,cdc.go:612,cdc.go:124,cluster.go:1582,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1163702-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		   0.0 newOrder
		  23m47s     5241            0.0            1.6      0.0      0.0      0.0      0.0 orderStatus
		  23m47s     5241            0.0           16.5      0.0      0.0      0.0      0.0 payment
		  23m47s     5241            0.0            1.6      0.0      0.0      0.0      0.0 stockLevel
		E190306 07:43:37.930046 1 workload/cli/run.go:419  error in newOrder: dial tcp 10.142.0.101:26257: connect: connection refused
		  23m48s     5273            0.0            1.6      0.0      0.0      0.0      0.0 delivery
		  23m48s     5273            0.0           16.9      0.0      0.0      0.0      0.0 newOrder
		  23m48s     5273            0.0            1.6      0.0      0.0      0.0      0.0 orderStatus
		  23m48s     5273            0.0           16.5      0.0      0.0      0.0      0.0 payment
		  23m48s     5273            0.0            1.6      0.0      0.0      0.0      0.0 stockLevel
		E190306 07:43:38.995501 1 workload/cli/run.go:419  error in stockLevel: dial tcp 10.142.0.101:26257: connect: connection refused
		: signal: killed
	cluster.go:1603,cdc.go:212,cdc.go:432,test.go:1214: unexpected node event: 1: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/57e825a7940495b67e0cc7213a5fabc24e12be0e

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1176948&tab=buildLog

```
The test failed on master:
	cluster.go:1251,cdc.go:625,cdc.go:125,cluster.go:1589,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1176948-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		    1.7      0.0      0.0      0.0      0.0 delivery
		  20m25s     4216            0.0           17.1      0.0      0.0      0.0      0.0 newOrder
		  20m25s     4216            0.0            1.7      0.0      0.0      0.0      0.0 orderStatus
		  20m25s     4216            0.0           16.6      0.0      0.0      0.0      0.0 payment
		  20m25s     4216            0.0            1.6      0.0      0.0      0.0      0.0 stockLevel
		E190314 06:41:31.560174 1 workload/cli/run.go:419  error in payment: dial tcp 10.142.0.115:26257: connect: connection refused
		  20m26s     4233            0.0            1.7      0.0      0.0      0.0      0.0 delivery
		  20m26s     4233            0.0           17.1      0.0      0.0      0.0      0.0 newOrder
		  20m26s     4233            0.0            1.7      0.0      0.0      0.0      0.0 orderStatus
		  20m26s     4233            0.0           16.6      0.0      0.0      0.0      0.0 payment
		  20m26s     4233            0.0            1.6      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1610,cdc.go:213,cdc.go:433,test.go:1214: unexpected node event: 2: dead

```

cc @nvanbenschoten / @danhhz 

```
190314 06:41:18.513779 470 server/status/runtime.go:464  [n2] runtime stats: 5.0 GiB RSS, 1203 goroutines, 184 MiB/344 MiB/675 MiB GO alloc/idle/total, 3.6 GiB/4.4 GiB CGO alloc/total, 49900.9 CGO/sec, 82.3/21.8 %(u/s)time, 0.0 %gc (4x), 7.2 MiB/78 MiB (r/w)net
I190314 06:41:21.153683 322988 storage/store_snapshot.go:818  [n2,raftsnapshot,s2,r357/2:/Table/56/1/7{5/6/-4â€¦-7/6/-2â€¦}] streamed snapshot to (n1,s1):1: kv pairs: 588072, log entries: 74, rate-limit: 8.0 MiB/sec, 5.34s
I190314 06:41:28.516211 470 server/status/runtime.go:464  [n2] runtime stats: 5.0 GiB RSS, 1208 goroutines, 301 MiB/302 MiB/675 MiB GO alloc/idle/total, 3.6 GiB/4.4 GiB CGO alloc/total, 43185.3 CGO/sec, 72.4/23.3 %(u/s)time, 0.0 %gc (2x), 7.5 MiB/35 MiB (r/w)net
I190314 06:41:28.755717 326561 storage/store_snapshot.go:775  [n2,raftsnapshot,s2,r30/2:/Table/53/2{-/37/10/â€¦}] sending Raft snapshot 4da2c536 at applied index 16358
E190314 06:41:28.766573 309986 util/log/crash_reporting.go:202  [n2,s2,r28/2:/Table/53/1/{39/6/-â€¦-79/1/-â€¦},rangefeed] a panic has occurred!
panic: resolved timestamp regression, was 1552545674.481467572,0, recomputed as 1552545640.548937542,0 [recovered]
	panic: resolved timestamp regression, was 1552545674.481467572,0, recomputed as 1552545640.548937542,0

goroutine 309986 [running]:
panic(0x2d7c2e0, 0xc00678ab30)
	/usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc0195c9520 sp=0xc0195c9490 pc=0x72a5bb
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000942000, 0x39db560, 0xc008899dd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x11f fp=0xc0195c9580 sp=0xc0195c9520 pc=0x108479f
runtime.call32(0x0, 0x33dc130, 0xc00dcf7430, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc0195c95b0 sp=0xc0195c9580 pc=0x75919b
panic(0x2d7c2e0, 0xc00678ab30)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc0195c9640 sp=0xc0195c95b0 pc=0x72a4a9
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).recompute(0xc007307880, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:189 +0x236 fp=0xc0195c96c8 sp=0xc0195c9640 pc=0x18bacb6
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).assertNoChange(0xc007307880)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:200 +0x57 fp=0xc0195c9760 sp=0xc0195c96c8 pc=0x18bad37
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).ConsumeLogicalOp(0xc007307880, 0x0, 0x0, 0x0, 0xc01ddc4f60, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:134 +0x89 fp=0xc0195c97b0 sp=0xc0195c9760 pc=0x18ba559
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeLogicalOps(0xc0073077c0, 0x39db4a0, 0xc00bb05d80, 0xc00fb062d0, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:535 +0x21b fp=0xc0195c98a8 sp=0xc0195c97b0 pc=0x18b771b
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeEvent(0xc0073077c0, 0x39db4a0, 0xc00bb05d80, 0xc00fb062d0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:487 +0x251 fp=0xc0195c9918 sp=0xc0195c98a8 pc=0x18b74d1
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).Start.func1(0x39db4a0, 0xc00bb05d80)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:245 +0x9bd fp=0xc0195c9f88 sp=0xc0195c9918 pc=0x18bd85d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc018bf1b80, 0xc000942000, 0xc008899d40)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0xe1 fp=0xc0195c9fc8 sp=0xc0195c9f88 pc=0x1086921
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc0195c9fd0 sp=0xc0195c9fc8 pc=0x75aeb1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:193 +0xa8
```

I just saw this on a run of cdc/tpcc-1000 I was doing with #35745. @ajwerner any chance this could be #35130?

```
E190314 19:02:30.978867 960980 util/log/crash_reporting.go:202  [n1,s1,r317/1:/Table/60/1/"{<\xad\…-@\x8c\…},rangefeed] a panic has occurred!
panic: resolved timestamp regression, was 1552590117.199945509,0, recomputed as 1552590102.203202946,1 [recovered]
        panic: resolved timestamp regression, was 1552590117.199945509,0, recomputed as 1552590102.203202946,1

goroutine 960980 [running]:
panic(0x2d79ae0, 0xc03b6d54d0)
        /usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc01a3b9520 sp=0xc01a3b9490 pc=0x72a5bb
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0001e06c0, 0x39d8180, 0xc015da5860)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x11f fp=0xc01a3b9580 sp=0xc01a3b9520 pc=0x108452f
runtime.call32(0x0, 0x33d92b0, 0xc025c1db60, 0x1800000018)
        /usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc01a3b95b0 sp=0xc01a3b9580 pc=0x75919b
panic(0x2d79ae0, 0xc03b6d54d0)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc01a3b9640 sp=0xc01a3b95b0 pc=0x72a4a9
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).recompute(0xc012b10ac0, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:189 +0x236 fp=0xc01a3b96c8 sp=0xc01a3b9640 pc=0x18ba476
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).assertNoChange(0xc012b10ac0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:200 +0x57 fp=0xc01a3b9760 sp=0xc01a3b96c8 pc=0x18ba4f7
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).ConsumeLogicalOp(0xc012b10ac0, 0x0, 0x0, 0x0, 0xc035c5dc80, 0x0, 0x0, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:134 +0x89 fp=0xc01a3b97b0 sp=0xc01a3b9760 pc=0x18b9d19
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeLogicalOps(0xc012b10a00, 0x39d80c0, 0xc0273b2400, 0xc03f2b0900, 0x9, 0x10)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:535 +0x21b fp=0xc01a3b98a8 sp=0xc01a3b97b0 pc=0x18b6edb
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeEvent(0xc012b10a00, 0x39d80c0, 0xc0273b2400, 0xc03f2b0900, 0x9, 0x10, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:487 +0x251 fp=0xc01a3b9918 sp=0xc01a3b98a8 pc=0x18b6c91
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).Start.func1(0x39d80c0, 0xc0273b2400)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:245 +0x9bd fp=0xc01a3b9f88 sp=0xc01a3b9918 pc=0x18bd01d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc02e787a30, 0xc0001e06c0, 0xc015da5800)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0xe1 fp=0xc01a3b9fc8 sp=0xc01a3b9f88 pc=0x10866b1
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc01a3b9fd0 sp=0xc01a3b9fc8 pc=0x75aeb1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:193 +0xa8
```

SHA: https://github.com/cockroachdb/cockroach/commits/70e3468e7ed5fb30b10eaaf972acbb0f16099d01

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/crdb-chaos/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1178890&tab=buildLog

```
The test failed on master:
	cluster.go:1267,cdc.go:625,cdc.go:125,cluster.go:1605,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1178890-cdc-crdb-chaos-rangefeed-true:4 -- ./workload run tpcc --warehouses=100 --duration=30m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		    1.0      0.0      0.0      0.0      0.0 delivery
		   7m47s     1630            0.0           10.9      0.0      0.0      0.0      0.0 newOrder
		   7m47s     1630            0.0            1.1      0.0      0.0      0.0      0.0 orderStatus
		   7m47s     1630            0.0           10.4      0.0      0.0      0.0      0.0 payment
		   7m47s     1630            0.0            0.7      0.0      0.0      0.0      0.0 stockLevel
		   7m48s     1651            0.0            1.0      0.0      0.0      0.0      0.0 delivery
		   7m48s     1651            0.0           10.8      0.0      0.0      0.0      0.0 newOrder
		   7m48s     1651            0.0            1.1      0.0      0.0      0.0      0.0 orderStatus
		   7m48s     1651            0.0           10.4      0.0      0.0      0.0      0.0 payment
		   7m48s     1651            0.0            0.7      0.0      0.0      0.0      0.0 stockLevel
		E190315 06:26:06.344892 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.127:26257: connect: connection refused
		: signal: killed
	cluster.go:1626,cdc.go:213,cdc.go:433,test.go:1214: unexpected node event: 2: dead

```

Here it goes again, but looks like @nvanbenschoten already figured this out: #35777

```
E190315 06:26:02.753584 22498 util/log/crash_reporting.go:202  [n2,s2,r78/3:/Table/53/1/"\x{b8wz\â€¦-e0i\xâ€¦},rangefeed] a panic has occurred!
panic: resolved timestamp regression, was 1552631155.802869173,0, recomputed as 1552631153.637038233,0 [recovered]
	panic: resolved timestamp regression, was 1552631155.802869173,0, recomputed as 1552631153.637038233,0

goroutine 22498 [running]:
```

--- a/pkg/storage/rangefeed/resolved_timestamp.go
+++ b/pkg/storage/rangefeed/resolved_timestamp.go
@@ -155,18 +155,11 @@ func (rts *resolvedTimestamp) consumeLogicalOp(op enginepb.MVCCLogicalOp) bool {
 		return rts.intentQ.DecrRef(t.TxnID, t.Timestamp)
 
 	case *enginepb.MVCCAbortIntentOp:
-		// If the resolved timestamp has been initialized then we can remove the
-		// txn from the queue immediately after we observe that it has been
-		// aborted. However, if the resolved timestamp has not yet been
-		// initialized then we decrement from the txn's reference count. This
-		// permits the refcount to drop below 0, which is important to handle
-		// correctly when events may be out of order. For instance, before the
-		// resolved timestamp is initialized, it's possible that we see an
-		// intent get aborted before we see it in the first place. If we didn't
-		// let the refcount drop below 0, we would risk leaking a reference.
-		if rts.IsInit() {
-			return rts.intentQ.Del(t.TxnID)
-		}
+		// An aborted intent does not necessarily indicate an aborted
+		// transaction. An AbortIntent operation can be the result of an intent
+		// that was written only in an earlier epoch being resolved after its
+		// transaction committed in a later epoch. Don't make any assumptions
+		// about the transaction other than to decrement its reference count.
 		return rts.intentQ.DecrRef(t.TxnID, hlc.Timestamp{})
 
 	default:
@@ -419,24 +412,6 @@ func (uiq *unresolvedIntentQueue) updateTxn(
 		return wasMin
 	}
 	return false
-}
-
-// Del removes the transaction from the queue. It returns whether the update had
-// an effect on the oldest transaction in the queue.
-func (uiq *unresolvedIntentQueue) Del(txnID uuid.UUID) bool {
-	txn, ok := uiq.txns[txnID]
-	if !ok {
-		// Unknown txn.
-		return false
-	}
-
-	// Will deleting the txn advance the queue's earliest timestamp?
-	wasMin := txn.index == 0
-
-	// Remove txn from the queue.
-	delete(uiq.txns, txn.txnID)
-	heap.Remove(&uiq.minHeap, txn.index)
-	return wasMin
 }
 
 // AllowNegRefCount instruts the unresolvedIntentQueue on whether or not to

--- a/pkg/storage/rangefeed/resolved_timestamp_test.go
+++ b/pkg/storage/rangefeed/resolved_timestamp_test.go
@@ -68,12 +68,6 @@ func TestUnresolvedIntentQueue(t *testing.T) {
 	require.False(t, adv)
 	require.Equal(t, 2, uiq.Len())
 
-	// Delete a non-existent txn.
-	txn4 := uuid.MakeV4()
-	adv = uiq.Del(txn4)
-	require.False(t, adv)
-	require.Equal(t, 2, uiq.Len())
-
 	// Update txn1 with a smaller timestamp.
 	adv = uiq.UpdateTS(txn1, hlc.Timestamp{WallTime: 0})
 	require.False(t, adv)
@@ -130,13 +124,13 @@ func TestUnresolvedIntentQueue(t *testing.T) {
 	require.Equal(t, 1, uiq.txns[txn1].refCount)
 	require.Equal(t, newTxn1TS, uiq.txns[txn1].timestamp)
 
-	// Add new txn at much higher timestamp. Immediately delete.
-	txn5 := uuid.MakeV4()
-	adv = uiq.IncRef(txn5, nil, hlc.Timestamp{WallTime: 10})
+	// Add new txn at much higher timestamp. Immediately decrement ref count.
+	txn4 := uuid.MakeV4()
+	adv = uiq.IncRef(txn4, nil, hlc.Timestamp{WallTime: 10})
 	require.False(t, adv)
 	require.Equal(t, 3, uiq.Len())
 	require.Equal(t, txn2, uiq.Oldest().txnID)
-	adv = uiq.Del(txn5)
+	adv = uiq.DecrRef(txn4, hlc.Timestamp{WallTime: 10})
 	require.False(t, adv)
 	require.Equal(t, 2, uiq.Len())
 
@@ -153,19 +147,19 @@ func TestUnresolvedIntentQueue(t *testing.T) {
 	require.True(t, adv)
 	require.Equal(t, 0, uiq.Len())
 
-	// Add new txn. Immediately delete. Should be empty again.
-	txn6 := uuid.MakeV4()
-	adv = uiq.IncRef(txn6, nil, hlc.Timestamp{WallTime: 20})
+	// Add new txn. Immediately decrement ref count. Should be empty again.
+	txn5 := uuid.MakeV4()
+	adv = uiq.IncRef(txn5, nil, hlc.Timestamp{WallTime: 20})
 	require.False(t, adv)
 	require.Equal(t, 1, uiq.Len())
-	require.Equal(t, txn6, uiq.Oldest().txnID)
-	adv = uiq.Del(txn6)
+	require.Equal(t, txn5, uiq.Oldest().txnID)
+	adv = uiq.DecrRef(txn5, hlc.Timestamp{})
 	require.True(t, adv)
 	require.Equal(t, 0, uiq.Len())
 
 	// Instruct the queue to disallow negative ref counts.
 	uiq.AllowNegRefCount(false)
-	require.Panics(t, func() { uiq.DecrRef(txn6, hlc.Timestamp{}) })
+	require.Panics(t, func() { uiq.DecrRef(txn5, hlc.Timestamp{}) })
 	require.Equal(t, 0, uiq.Len())
 }
 
@@ -264,6 +258,8 @@ func TestResolvedTimestamp(t *testing.T) {
 	// Third transaction aborted. No effect.
 	fwd = rts.ConsumeLogicalOp(abortIntentOp(txn3))
 	require.False(t, fwd)
+	fwd = rts.ConsumeLogicalOp(abortIntentOp(txn3))
+	require.False(t, fwd)
 	require.Equal(t, hlc.Timestamp{WallTime: 24}, rts.Get())
 
 	// Fourth transaction at higher timestamp. No effect.
@@ -314,7 +310,19 @@ func TestResolvedTimestamp(t *testing.T) {
 	require.True(t, fwd)
 	require.Equal(t, hlc.Timestamp{WallTime: 45}, rts.Get())
 
-	// Fifth transaction aborted. Resolved timestamp moves to closed timestamp.
+	// Fifth transaction bumps epoch and re-writes one of its intents. Resolved
+	// timestamp moves to the new transaction timestamp.
+	fwd = rts.ConsumeLogicalOp(updateIntentOp(txn5, hlc.Timestamp{WallTime: 47}))
+	require.True(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 46}, rts.Get())
+
+	// Fifth transaction committed, but only one of its intents was written in
+	// its final epoch. Resolved timestamp moves forward after observing the
+	// first intent committing at a higher timestamp and moves to the closed
+	// timestamp after observing the second intent aborting.
+	fwd = rts.ConsumeLogicalOp(commitIntentOp(txn5, hlc.Timestamp{WallTime: 49}))
+	require.True(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 48}, rts.Get())
 	fwd = rts.ConsumeLogicalOp(abortIntentOp(txn5))
 	require.True(t, fwd)
 	require.Equal(t, hlc.Timestamp{WallTime: 50}, rts.Get())

